Mathlib/RingTheory/Flat/Localization.lean

```diff
@@ -48,57 +48,68 @@ theorem flat_iff_of_isLocalization : Flat S M ↔ Flat R M :=
   have := IsLocalization.flat S p
   ⟨fun _ ↦ .trans R S M, fun _ ↦ .of_isLocalizedModule S p .id⟩
 
-private lemma aux (I : Ideal R) (s : Submonoid R) :
-    have hM := isBaseChange s (Localization s) (mkLinearMap s M)
-    have hIM := isBaseChange s (Localization s) (mkLinearMap s (I ⊗[R] M))
-    let e := (hIM.equiv.restrictScalars R).symm ≪≫ₗ
-      leftComm _ _ _ _ ≪≫ₗ (hM.equiv.restrictScalars R).lTensor I
-    LocalizedModule.map s (TensorProduct.lift <| lsmul R M ∘ₗ I.subtype) =
-      TensorProduct.lift (lsmul R (LocalizedModule s M) ∘ₗ I.subtype) ∘ₗ e.toLinearMap := by
-  refine linearMap_ext s (mkLinearMap s _) (mkLinearMap s _) ?_
-  ext m i
-  rw [AlgebraTensorModule.curry_apply, curry_apply, restrictScalars_apply, LinearMap.comp_apply,
-    restrictScalars_apply, mkLinearMap_apply, AlgebraTensorModule.curry_apply, curry_apply,
-    restrictScalars_apply]
-  simpa [-mkLinearMap_apply, IsBaseChange.equiv_symm_apply, IsBaseChange.equiv_tmul] using
-    (mkLinearMap_apply _ _ _).symm.trans (map_smul (mkLinearMap s M) _ _)
-
-variable (Mₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
-  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Mₚ P)]
-  [∀ (P : Ideal R) [P.IsMaximal], Module R (Mₚ P)]
-  (f : ∀ (P : Ideal R) [P.IsMaximal], M →ₗ[R] Mₚ P)
-  [∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (f P)]
+noncomputable section
+
+variable {S} {I : Ideal R} {s : Submonoid S}
+variable {Mₛ : Type*} [AddCommGroup Mₛ] [Module R Mₛ] [Module S Mₛ] [IsScalarTower R S Mₛ]
+variable (f : M →ₗ[S] Mₛ) [IsLocalizedModule s f]
+local notation "mI" => AlgebraTensorModule.map f (LinearMap.id (R := R) (M := I))
+local notation "mR" => AlgebraTensorModule.map f (LinearMap.id (R := R) (M := R))
+
+private lemma aux :
+    ⇑(LocalizedModule.map s <| AlgebraTensorModule.lTensor S M I.subtype) =
+      (iso s mR).symm ∘ₗ AlgebraTensorModule.lTensor S Mₛ I.subtype ∘ₗ iso s mI := by
+  rw [← LinearMap.coe_restrictScalars S]
+  apply congr_arg
+  rw [LinearEquiv.eq_toLinearMap_symm_comp, ← LinearEquiv.comp_toLinearMap_symm_eq]
+  refine linearMap_ext s mI mR ?_
+  rw [LinearMap.comp_assoc, iso_symm_comp]
+  ext; simp
```

Comment:
I think this lemma could be public and be put under `isLocalizedModule_tensorProduct_map`. It holds for (Comm?)Semiring/AddCommMonoid and arbitrary linear maps in place of `I.subtype`.

Reply:
And should we define `AlgebraTensorModule.rTensor` to be `AlgebraTensorModule.map _ LinearMap.id` (parallel to `LinearMap.rTensor`) and develop some API around it?

Reply:
I added `AlgebraTensorModule.lTensor` recently, so I am in favor of also adding `AlgebraTensorModule.rTensor`.

Reply:
Okay I refactored a bit: I added `IsLocalizedModule.map_lTensor` and `LocalizedModule.restrictScalars_map_eq` (the latter is unused in this PR, but it is your generalized `aux` lemma), removed `aux` and now use `injective_of_isLocalized_{span,maximal}` directly instead of going via `IsLocalizedModule.iso`.